### PR TITLE
Structured labelled arrays

### PR DESCRIPTION
**Pull request by ericmjl**
This PR adds three functions allow a user to generate _labelled_ adjacency tensors and _labelled_ node feature matrices from a graph.

### CODE REVIEW

**Comment by dschult** [comment]
I'm a little confused by all the pep8 stuff.  Shouldn't running black once fix all that? 
Is it true that black doesn't mess with comments and doc_strings but that flake8 and pycodestyle do?
How can we make this easier for submitters?

File: networkx/convert_matrix.py [hunk 1]
```diff
@@ -397,14 +400,16 @@ def from_pandas_edgelist(
...
+            c for c in df.columns if c not in reserved_columns
+        ]
     elif isinstance(edge_attr, (list, tuple)):
         attr_col_headings = edge_attr
     else:
         attr_col_headings = [edge_attr]
     if len(attr_col_headings) == 0:
         raise nx.NetworkXError(
-            f"Invalid edge_attr argument: No columns found with name: {attr_col_headings}"
+            f"Invalid edge_attr argument: No columns found with name: {attr_col_headings}"  # noqa: E501
```
Comment (dschult) [inline networkx/convert_matrix.py:412]: Can we take out ```noqa``` stuff, since we're moving to relying on black?
I don't know the answer, but it'd be nice not to have distractions in the code.

File: networkx/convert_matrix.py [hunk 2]
```diff
@@ -1360,3 +1374,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+    Return a pandas DataFrame representation of nodes and their metadata.
+
+    This function accepts a graph and a collection of user-defined functions
+    and returns a pandas DataFrame in which the nodes' are the DataFrame index,
+    the node metadata keys are column names,
+    and node metadata values are the values in the DataFrame.
+
+    ``funcs`` is an iterable of callables whose signature is
+
+        f(n, d) -> pd.Series
```
Comment (dschult) [inline networkx/convert_matrix.py:1393]: Why should this be returning a ```pd.Series```? Isn't it just computing a single value? I don't know a huge amount about ```pd.Series```.  Is it better to create a bunch of Series and then combine them? Or is it better to collect the values into a single Series?  

Another way to ask this question:  Would it be better to move the ```pd.Series``` part of these functions outside the function? Then the user's function could focus on creating the value rather than packaging it.
Seems like the calling function knows the node name and can construct the Series for the user.

File: networkx/convert_matrix.py [hunk 3]
```diff
@@ -1360,3 +1374,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+
+    In other words, it accepts a (node, metadata dictionary) pair,
```
Comment (dschult) [inline networkx/convert_matrix.py:1395]: ```metadata dictionary``` is called ```nodeattr dict``` elsewhere... or more formally ```node attribute dict```

File: networkx/convert_matrix.py (hunk 1 above)
Comment (jarrodmillman) [reply networkx/convert_matrix.py:412]: If you replace
```
f"Invalid edge_attr argument: No columns found with name: {attr_col_headings}"
```
with
```
f"Invalid edge_attr: No columns found with name: {attr_col_headings}"
```
it won't get wrapped.

**Comment by jarrodmillman** [comment]
> I'm a little confused by all the pep8 stuff. Shouldn't running black once fix all that?
> Is it true that black doesn't mess with comments and doc_strings but that flake8 and pycodestyle do?
> How can we make this easier for submitters?

@dschult We need to set up black to run as part of the CI and remove pep8speaks.  @rossbar started looking into how to do this and has a reasonable idea of what to do (I think).

File: examples/convert/graph_tensors.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+"""
+=========================================================================
+Generating Node Feature Matrices and Diffusion Matrix Tensors from Graphs
+=========================================================================
+
+In this example, we show how two things:
+
+1. How to obtain a node dataframe from NetworkX graphs
+using custom user-defined functions,
+2. How to obtain an adjacency tensor from NetworkX graphs,
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply examples/convert/graph_tensors.py]: Will be pushing up an explanation of a tensor, and how they relate to numpy arrays.

File: examples/convert/graph_tensors.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+using custom user-defined functions.
+
+This allows you to generate structured dataframes and xarray dataarrays
+in a highly customized fashion.
+The resulting data containers are highly compatible
+with the rest of the PyData ecosystem,
+and can easily be converted into NumPy arrays/tensors
+to be fed into deep learning models.
+"""
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply examples/convert/graph_tensors.py]: This is a great suggestion, thanks @dschult!

File: networkx/conftest.py [hunk 6]
```diff
@@ -38,10 +38,14 @@ def set_warnings():
         message="literal_destringizer is deprecated*",
     )
     warnings.filterwarnings(
-        "ignore", category=DeprecationWarning, message="is_string_like is deprecated*"
+        "ignore",
+        category=DeprecationWarning,
+        message="is_string_like is deprecated*",
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply networkx/conftest.py:43]: @dschult I tried disabling my vscode settings (which clobber over the project settings), but that didn't automatically reformat things like the original, so I've manually reset it.

File: networkx/convert_matrix.py (hunk 1 above)
Comment (ericmjl) [reply networkx/convert_matrix.py:412]: That's true. For consistency, I've modified all of the other lines in the same function that carry the term `argument` in the error message.

File: examples/convert/graph_tensors.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    A = np.linalg.matrix_power(A, n)
+    return nx.format_adjacency(G, A, name=f"adjacency_{n}")
+
+
+def laplacian(G):
+    A = np.asarray(nx.laplacian_matrix(G).todense())
+    return nx.format_adjacency(G, A, name="laplacian_matrix")
+
+
+G = make_graph()
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply examples/convert/graph_tensors.py]: I was practicing making examples idempotent 😄, especially since the example is quite similar in style to the test. Hope you don't mind this?

File: networkx/convert_matrix.py (hunk 2 above)
Comment (ericmjl) [reply networkx/convert_matrix.py:1382]: It's a good point, @dschult. I think the reason why I've structured the requirement this way is to nudge end-users to using the community-wide standard and idiomatic PyData data structures. Something I've learned over the past few years is that targeting the "right" data structures can greatly simplify the kind of code we write, and I think tooling that explicitly nudges users towards the "right" data structures for a problem can help encourage better code writing altogether.

A series is nice for other reasons too. For example, it behaves like an upgraded dictionary; a collection of Series becomes a DataFrame. Underneath the hood, users can define functions that compute a whole bunch of values at one shot, such as in the `plot_graph_tensors.py` file where `node_metadata(n, d)` automagically converts all node metadata into a pandas Series, thereby giving a whole lot of values at one shot, or they can compute one value only (like in `ten_times_number(n, d)`).

File: networkx/convert_matrix.py (hunk 3 above)
Comment (ericmjl) [reply networkx/convert_matrix.py:1384]: Let me change that.

File: networkx/convert_matrix.py (hunk 2 above)
Comment (dschult) [reply networkx/convert_matrix.py:1383]: Interesting... I think a light clicked from the ```node_metadata(n, d)``` example where the Series contains many columns of data for one node. In my lingo (that os probably not quite correct) the pd.Series we create in this function is a "row Series" for that node... not a "column Series" (many nodes with one computed value).  I have typically thought of DataFrames as collections of "column Series", but it is very nice to realize that they are also collections of "row Series" as well.    Thanks!

Maybe you could change the wording below (in line 1388) to make explicit the idea that the pandas Series created in a function can consist of multiple values. The collection ```funcs``` stacks these Series to make a record for each node. [change the words to make them match the usual idioms for pandas/xarray.]  I suppose this might be made clear in the example by including ```"ycoord": d["y_coord"]``` in the dict as well... but I'm not sure what is the best presentation.

File: networkx/convert_matrix.py [hunk 8]
```diff
@@ -1360,3 +1364,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+
+    .. code-block:: python
+
+        def x_vec(n: Hashable, d: Dict[Hashable, Any]) -> pd.Series:
+            return pd.Series({"x_coord": d["x_coord"]}, name=n)
+
+    One fairly strong assumption is that each func
+    has all the information it needs to act
+    stored on the attribute dictionary,
+    and does not take in any other keyword arguments.
```
Comment (dschult) [inline networkx/convert_matrix.py:1400]:
```suggestion
    One fairly strong assumption is that the attribute dictionary
    has all the information it needs to act
    and does not need any other keyword arguments.
```

File: networkx/convert_matrix.py [hunk 9]
```diff
@@ -1360,3 +1364,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+    .. code-block:: python
+
+        from functools import partial
+        def get_molweight(n, d, mw_dict):
+            return pd.Series({"mw": mw_dict[d["amino_acid"]]}, name=n)
+
+        mw_dict = {"PHE": 165, "GLY": 75, ...}
+        get_molweight_func = partial(get_molweight, mw_dict=mw_dict)
+
+        generate_feature_dataframe(G, [get_molweight_func])
```
Comment (dschult) [inline networkx/convert_matrix.py:1418]: Can ```lambda``` functions work too?  
Like  ```generate_feature_dataframe(G, [lambda n, d:get_molweight(n, d, mw_dict)])```

I assume so... is it worth saying?

File: networkx/convert_matrix.py [hunk 10]
```diff
@@ -1360,3 +1364,174 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+
+    The series that is returned from each function
+    need not only contain one key-value pair.
+    You can have two or more, and that's completely fine;
+    each key becomes a column in the resulting dataframe.
+
+    A key design choice: We default to returning DataFrames,
+    to make inspecting the data easy,
+    For consumption in tensor libraries,
+    you can extract the DataFrame values after the fact.
```
Comment (dschult) [inline networkx/convert_matrix.py:1433]: What does ```consumption in tensor libraries`` refer to?

Comment (ericmjl) [reply networkx/convert_matrix.py:1432]: Updating!

File: networkx/convert_matrix.py (hunk 9 above)
Comment (ericmjl) [reply networkx/convert_matrix.py:1417]: Yes! I've used the same pattern too 😄.

**Comment by jarrodmillman** [comment]
The example doesn't render correctly:
https://324-890377-gh.circle-artifacts.com/0/doc/build/html/auto_examples/convert/plot_graph_tensors.html#sphx-glr-auto-examples-convert-plot-graph-tensors-py

I think changing
```
An overview of the steps involved looks like this:

1. Create your graph with data attributes
1. Define functions to extract data.
    - For node DataFrames functions have signature:
    f(node, datadict) -> pd.Series.
    - For adjacency DataArray functions have signature: f(G) -> xr.DataArray.
    Use ``format_adjacency()`` to ease creation of the DataArray.
1. Call the relevant generate function
(``generate_node_dataframe`` or ``generate_adjacency_xarray``)

"""
```
to
```
An overview of the steps involved looks like this:

1. Create your graph with data attributes
2. Define functions to extract data.

    - For node DataFrames functions have signature:
      f(node, datadict) -> pd.Series.
    - For adjacency DataArray functions have signature: f(G) -> xr.DataArray.
      Use ``format_adjacency()`` to ease creation of the DataArray.

3. Call the relevant generate function
   (``generate_node_dataframe`` or ``generate_adjacency_xarray``)
"""
```
should fix it.  That is,

1. use different numbers (so it makes sense to people who read the text file).
2. use spaces to mark the beginning and end of the bullet list.
3. indent text to keep it in the same block (e.g., in the bullet list and item 3).

You can also remove the extra space at the end of the docstring.

I would also use a shorter title and a more descriptive first sentence.  The short title is the image title in the gallery.   I would use 3-4 words so the title in the gallery isn't super long.  I would add a descriptive first sentence, since it is what is displayed when you hover over the image in the gallery (it currently says "In this example, we show how two things:").  Maybe make the current title the first sentence and add a shorter less descriptive title.

If possible it would be nice to draw a figure.  The figure is used in the gallery and makes it more visually interesting.  Otherwise, it just displays a default image.  You may want to look at
- https://xarray.pydata.org/en/stable/plotting.html

for ideas.  I.e., it would be nice if the plot used the fact that you have an xarray.

Should "f(node, datadict) -> pd.Series" be ``f(node, datadict) -> pd.Series`` (i.e., put in double backquotes)?  And "f(G) -> xr.DataArray" be ``f(G) -> xr.DataArray``?

**Comment by ericmjl** [comment]
Thanks @jarrodmillman! Yes, I was sticking with Markdown conventions, and forgot that we're using Sphinx here.

I'm wondering, do you know if it's possible to display the HTML repr of an xarray object in sphinx? That would be the most rad. But if not, I have an idea of what to put in -- possibly just a heatmap of sorts.

**Comment by ericmjl** [comment]
@jarrodmillman @dschult looks like all tests pass, and the built example looks pretty cool too! :smile:

File: networkx/convert_matrix.py (hunk 2 above)
Comment (ericmjl) [reply networkx/convert_matrix.py:1383]: I think this can be marked as resolved.

File: networkx/convert_matrix.py [hunk 11]
```diff
@@ -1360,3 +1364,192 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
     return from_numpy_matrix(
         A, parallel_edges=parallel_edges, create_using=create_using
     )
+
+
+def generate_node_dataframe(
```
Comment (dschult) [inline networkx/convert_matrix.py:1369]: The naming scheme for the rest of this module is to use ```to``` instead of ```generate```. 
 For example: Is ```to_node_dataframe``` a poor choice for this function?
Even if that doesn't work, I would encourage us not to use ```generate_XXX``` as we already have Graph generators and that name has been problematic because of Python's use of the word ```generator```. So I worry that this naming choice might be problematic too. 

I think we should also decide how to name functions that do this very thing but create a numpy array instead of a pandas dataframe. I believe that originally this PR had a keyword option to enable that... but it actually constructed the pandas dataframe and then converted to numpy array in that case.  Have you thought about/tried a version that constructs the numpy.array and then adds the names to the rows and columns when it converts to a dataframe? Same sort of idea when you construct a list of Series building it into the dataframe.  Would it be better to construct the dataframe directly?

I don't know when it is best to construct with lists/dicts, with numpy.arrays or with pandas dataframes or xarrays?  Do any of them present a way to input directly, or do they all come back to constructing the data as lists/dicts and then reading it into the object?

File: networkx/convert_matrix.py [hunk 12]
```diff
@@ -1360,3 +1364,192 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+    1. `name` is something that is unique amongst all names used
+    in the final adjacency tensor.
+
+    :param G: NetworkX-compatible Graph
+    :param adj: 2D numpy array
+    :param name: A unique name for the kind of adjacency matrix
+        being constructed.
+        Gets used in xarray as a coordinate in the "name" dimension.
+    :returns: An XArray DataArray of shape (n_nodes, n_nodes, 1)
+    """
```
Comment (dschult) [inline networkx/convert_matrix.py:1504]: Should these doc_strings be using the sphinx syntax for showing parameters in docstrings?

**Review by dschult — APPROVED** [review]
What about having a function ```from_node_dataframe```?

Anyway, I think once the name choice and sphinx syntax are worked out, this is ready for merging.

We can return in other PRs to address the questions about whether it is better to construct via lists or numpy.arrays or somehow directly into pandas and xarray. Sorry for the noise...

File: networkx/convert_matrix.py [hunk 13]
```diff
@@ -1360,3 +1364,192 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
     return from_numpy_matrix(
         A, parallel_edges=parallel_edges, create_using=create_using
     )
+
+
+def generate_node_dataframe(
+    G: nx.Graph,
+    funcs: List[Callable],
+):
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply networkx/convert_matrix.py:1372]: Cool, got it - thanks @dschult!

File: networkx/convert_matrix.py (hunk 11 above)
Comment (ericmjl) [reply networkx/convert_matrix.py]: > The naming scheme for the rest of this module is to use to instead of generate.
> For example: Is to_node_dataframe a poor choice for this function?

Not at all! I have changed the function name to use `to_`. :smile:

> Have you thought about/tried a version that constructs the numpy.array and then adds the names to the rows and columns when it converts to a dataframe? Same sort of idea when you construct a list of Series building it into the dataframe. Would it be better to construct the dataframe directly?

Syntactically, it's more compact to construct the DataFrame/Series first before converting to an array. It's more complex, I think, to construct the array and then tag on the metadata. Something about the latter makes me less at ease, in that it's easier in our codebase to mess up the labelling of the DataFrame by constructing of the array first. (After all, it's easy to get the shape right while also getting the semantic ordering/labelling of rows/columns wrong.) We externalize that level of complexity by calling on pandas' array-returning methods/attributes after-the-fact.

> I don't know when it is best to construct with lists/dicts, with numpy.arrays or with pandas dataframes or xarrays? Do any of them present a way to input directly, or do they all come back to constructing the data as lists/dicts and then reading it into the object?

The way I think about it is as follows: in a hierarchy of complexity, DataFrames are the "most complex", followed by arrays, followed by Python primitive data structures. Though they are the most complex, DataFrames also provide convenience methods to drop down to arrays and lists if a user so desires. Because it's easy for an end-user to do so, by either a `df[column].to_list()` for lists or `df.values` for arrays, I think Ross' previous comment on externalizing this piece to the end-user is a good rule of thumb to follow.

File: networkx/convert_matrix.py (hunk 12 above)
Comment (ericmjl) [reply networkx/convert_matrix.py]: Thanks for the catch, @dschult. I noticed it's NumPy-style elsewhere, so I've pushed up a change that changes the docstrings to NumPy-style.

**Comment by ericmjl** [comment]
@dschult to address the final point:

> What about having a function from_node_dataframe?

I like the idea! I think it could be left for another newcomer to contribute, since I think the pattern is similar:

```python
def from_node_dataframe(df):
    G = nx.Graph()
    for n, d in df.iterrows():
        G.add_node(n, **d)
    return G
```

Probably could be made more efficient, more flexible with custom funcs too, but I think it's worth fleshing out in a different PR, maybe with a design sketch via discussion on the issue tracker! I'll raise this up there.

**Comment by ericmjl** [comment]
Chiming in here for a gentle nudge on the PR. No pressure though, I do realize we're all busy people; just didn't want the thread to get dropped.

**Review by rossbar — COMMENTED** [review]
I took the liberty of doing a bit of rst reformatting & adding intersphinx links for pandas/xarray.

I really like the example! I think it's a great addition showing how to interoperate with other array/data libraries.

After taking a closer look at the functions - I'm not sure that some of the functions are worth the expanded API surface. For example, `to_adjacency_xarray` essentially boils down to a one-liner: `xr.concat([func(G) for func in funcs], dim="name")`. I'd be concerned about adding helper functions that don't really have much to do with NetworkX. Maybe this could be discussed further?

In summary, at this stage I'm:
 +1 on the new `plot_graph_tensors` example
 +1 on adding the `to_node_dataframe` function
 -0.5 on `to_adjacency_xarray` and `format_adjacency` functions
 -1 on adding a pyproject.toml (at least in this PR)

File: networkx/convert_matrix.py [hunk 14]
```diff
@@ -1360,3 +1364,220 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
...
+
+    each funcion in ``funcs`` should have the signature:
+
+        f(n, d) -> pd.Series
+
+    As an example:
+
+    .. code-block:: python
+
+        def x_vec(n: Hashable, d: Dict[Hashable, Any]) -> pd.Series:
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply networkx/convert_matrix.py:1406]: Yeah, the type annotations were a holdover from when I first developed the code.

File: networkx/convert_matrix.py [hunk 15]
```diff
@@ -384,7 +388,7 @@ def from_pandas_edgelist(
     AtlasView({'A': {'weight': 3, 'color': 'red'}, 'D': {'weight': 6, 'color': 'blue'}})
 
 
-    """
+    """  # noqa: E501
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply networkx/convert_matrix.py:391]: Possibly. My local flake8 tells me the line is too long. No big deal, I'll just default to using what the NetworkX CI shows. That said, providing a locally-available config file, I think, will be very helpful for developers. Waiting for the CI to start up only to find out that I missed one or two code formatting errors because of a mismatched config can build up frustration in the long-run. (For example, I always black configured in VSCode to format to 79 char line lengths, as that's optimal for using VSCode on a 15" laptop or larger to see two source files side-by-side, but this conflicts with some projects.)

**Comment by ericmjl** [comment]
Regarding `format_adjacency` and `to_adjacency_xarray`, let me see if I understand where you're coming from - if the function doesn't call on core NetworkX functions or objects, then it probably shouldn't be included in the `convert_matrix.py` library of functions, is that right? That's the only distinguishing mark of these two functions compared to `to_node_dataframe`. 

If so, then yes, there's an opportunity for discussion here! If I may, I'd like to lay out the arguments in favour of keeping the two functions in the library.

I think the two functions do belong inside NetworkX, because they are related to producing the array form(s) of a graph, particularly, the "diffusion matrices" that are used for linear algebra forms of message passing on a graph. Though they don't explicitly use much of the NetworkX API, they are conceptually tied to applied graph theory. Now, as we know from handling high dimensional arrays, even 3D ones get confusing quickly without names, so providing helper functions that construct named arrays (i.e. xarrays and dataframes!) can be super helpful. 

From a practical standpoint, if I take out `to_adjacency_xarray` and `format_adjacency`, then the `plot_graph_tensors` example needs to be heavily modified by adding `to_adjacency_xarray` and `format_adjacency` explicitly in there, as `plot_graph_tensors.py` relies on those two functions. I'd also end up copy/pasting those two functions into another library where it'd live (probably `graphein`, for which I'm working with the lead developer to target NetworkX graphs as the idiomatic data structure). So now we end up with substantial code duplication: one copy of the code lives in an example in NetworkX, and another copy of the code lives in graphein. 

As an aside, I also thought a bit about the API. It feels a bit too heavy right now. There's one experiment I did, inside a Jupyter notebook, taking `format_adjacency` and turning it into a decorator:

```python
from functools import wraps
import numpy as np
import xarray as xr
import functools

def format_adjacency(func):
    @wraps(func)
    def inner(G, *args, **kwargs):
        name = kwargs.pop("name", None)
        if name is None:
            name = func.__name__

        adj = func(G, *args, **kwargs)
        expected_shape = (len(G), len(G))
        if adj.shape != expected_shape:
            raise ValueError(
                "Adjacency matrix is not shaped correctly, "
                f"should be of shape {expected_shape}, "
                f"instead got shape {adj.shape}."
            )
        #### THE MAGIC HAPPENS HERE #####
        adj = np.expand_dims(adj, axis=-1)
        adj = xr.DataArray(adj)
        nodes = list(G.nodes())
        return xr.DataArray(
            adj,
            dims=["n1", "n2", "name"],
            coords={"n1": nodes, "n2": nodes, "name": [name]},
        )
    return inner
```

With this, we can write code that is much more lightweight:

```python
from functools import partial
from networkx import to_adjacency_xarray, format_adjacency
import networkx as nx
import numpy as np

G = nx.erdos_renyi_graph(n=30, p=0.3)

@format_adjacency
def adjacency(G):
    return nx.adjacency_matrix(G).todense()

@format_adjacency
def adjacency_power(G, n):
    a = nx.adjacency_matrix(G).todense()
    return np.linalg.matrix_power(a, n)

adjacency_2 = partial(adjacency_power, n=2, name="adj_2")
funcs = [
    adjacency,
    adjacency_2
]

da = to_adjacency_xarray(G, funcs)
```

Of course I'd have to document that `name` is a reserved keyword argument, or use a much more explicit keyword argument name. This way, end-users only have to worry about decorating their numpy-array-generating functions with one decorator. Makes usage much easier.

**Comment by rossbar** [comment]
> if the function doesn't call on core NetworkX functions or objects, then it probably shouldn't be included in the convert_matrix.py library of functions, is that right?

No, that's not the concern - since the `convert*` modules are related to converting graphs to other formats, this does indeed seem the appropriate place for such conversion functions.

My concern is more about the cost/benefit of expanding the API. The one that really stands out to me is the `to_adjacency_matrix` function. That essentially boils down to a one-liner: `my_adj_ary = xr.concat([func(G) for func in funcs], name="dim")`. My concern is this - is it worth adding specialized API to do something that is already achievable in a very "Pythonic" way? For this reason especially, it feels like the function belongs in the example rather than in the library because it demonstrates the right way to go about this, using the appropriate tools. `format_adjacency` has a similar feel: half the function is dedicated to error checking and the part that does the conversion essentially boils down to things that simply accomplished using the tools provided by the various libraries involved, e.g.

```python
nodes = list(G.nodes())
my_formatted_adj = xr.DataArray(
    adj[..., np.newaxis], 
    dims=["n1", "n2", "name"],
    coords={"n1": nodes, "n2": nodes, "name": [name]},
)
```

Adding the extra layer of indirection for this functionality feels to me like a bit of a violation of one of the zen principles: *There should be one—and preferably only one—obvious way to do it.*

> That's the only distinguishing mark of these two functions compared to to_node_dataframe.

IMO the difference with `to_node_dataframe` is that it performs a remapping of the node/attribute data, which is a more complicated procedure that feels more atomic.

I very well may be just be thinking too conservatively re: the expanded API. I don't at all doubt the utility of mapping graphs to named nd data structures!

**Comment by ericmjl** [comment]
> My concern is this - is it worth adding specialized API to do something that is already achievable in a very "Pythonic" way? For this reason especially, it feels like the function belongs in the example rather than in the library because it demonstrates the right way to go about this, using the appropriate tools. format_adjacency has a similar feel: half the function is dedicated to error checking and the part that does the conversion essentially boils down to things that simply accomplished using the tools provided by the various libraries involved

Got it! I see where you're coming from now.

There is a second perspective, wanted to get your thoughts here. The API surface area provides functions that return specialized data structures already. For example, there's functions that target numpy arrays and pandas dataframes. Would a function that returns xarray DataArrays fall in the same category? And if so, would the previous concerns you raised nullify having the expanded API still?

If the consensus is still "let's not keep the function in the library", then I can move it to graphein instead. You made a good point that the bulk of the code is checking, so there's minimal duplication if I just moved the "soul" of the functions into the examples.

**Comment by rossbar** [comment]
> The API surface area provides functions that return specialized data structures already.

This is a good point. Looking at the other functions that are already in `convert_matrix`, there are some that fit the "pass-through" description. For example, `to_numpy_matrix` is essentially: `return np.asmatrix(to_numpy_array(*args, **kwargs))`. I would vote to deprecate functions like this as well (thanks for the reminder!) for the same reason mentioned above. Most of the others functions in the module have more complex procedures that translate between the Graph structures and the "other" format (DataFrame, ndarray, etc.) - I think `to_node_dataframe` fits this bill - which IMO are worthwhile.

File: networkx/convert_matrix.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jarrodmillman) [reply networkx/convert_matrix.py:391]: NetworkX is using the default settings for black.  The benefit of using black is diminished if we are going each use different settings for black.  This PR changes line lengths for your setting of black.  If I edit this file later, black will reformat those lines to the default length.

**Comment by jarrodmillman** [comment]
I originally (mistakenly) thought this only added xarray as a dependency for the gallery example.  This is the first example of adding a new extra dependency since we adopted our new "policy" for extra dependencies: 
```
Default dependencies are listed in requirements/default.txt and extra (i.e., optional) dependencies
are listed in requirements/extra.txt. We don't often add new default and extra dependencies. If you
are considering adding code that has a dependency, you should first consider adding a gallery example.
Typically, new proposed dependencies would first be added as extra dependencies. Extra dependencies
should be easy to install on all platforms and widely-used. New default dependencies should be easy
to install on all platforms, widely-used in the community, and have demonstrated potential for
wide-spread use in NetworkX.
```

Could all of the code move to the example (as we recommend for new extra dependencies)?  If not, we should take the opportunity to update the policy to better explain our reasoning about when new dependencies can skip first being added as examples.

If we decide to make xarray an extra dependency, then xarray needs to be listed  in ``requirements/extra.txt`` not ``requirements/example.txt``.  We should also mention the addition of a new extra dependency in the release notes.

**Comment by jarrodmillman** [comment]
We already made some significant changes to our dependencies for the next release.   I don't think it will cause any issues, but it makes me a little hesitant to make more dependency changes in general for the next release.

One option would be to add this as an example for the next release and then add it as an extra dependency for the 3.0 release.  That might allow us to see if there are any unexpected issues.

This approach might be more desirable if there were a couple of additional dependencies that are added in the same way for the 3.0 release.  We should discuss whether there are other potential extra dependencies that we want to add before the 3.0 release.

**Comment by jarrodmillman** [comment]
@ericmjl This branch had gotten out-of-sync with master.  It was a bit of work to rebase, so I decided to just push the changes to your branch.  I was careful and I don't think I messed anything up.  But you may want to double-check I didn't make any errors.

**Comment by jarrodmillman** [comment]
I made xarray a default dependency so the tests would run, which is why the tests are failing now.  I am still not sure if this should be a default dependency or an extra one.  I would also prefer making this a gallery example first, but I don't feel strongly about that. In general, I think xarray is the type of thing it would make sense to add as a default or extra dependency.

I don't have much experience using xarray, so I am not sure this is a general solution or if there are other ways to do it.  It would be helpful if we could get feedback from other people using xarray and nx.  But there may not be a lot of people with more experience with this than Eric.

**Comment by dschult** [comment]
I'm looking at this PR again, and I'm getting more excited about its inclusion.

We haven't really made edge and node attributes first class entities in NetworkX. 
We allow them and use them, but functions like ```nx.get/set_node/edge_attributes```
are pretty minimal. The functions in this PR increase our set of tools for manipulating node and edge attributes.  

I wonder if they should be housed with ```get/set_node/edge_attributes``` instead of in ```convert```?
They aren't converting a graph. They are exporting attributes. I'd like to see functions to
import attributes (or at least document how to import them. But that doesn't have to be in this PR.

**Comment by dschult** [comment]
I think the funcs should have reasonable defaults -- like writing all attributes in the ```edge/node_attr_dicts```.
Then the ```to_node_dataframe``` output has columns corresponding to the attributes on the nodes.
And the ```to_adjacency_dataarray``` output has adjacency arrays for each attribute on the edges.

I'm pretty sure we could get a better name for ```format_adjacency```.  Perhaps ```adjacency_array_to_dataarray```  (or perhaps not) :) 
     
Also, using "n1" and "n2" and "name" for the 3 axes in the xarray could be made more specific:  
perhaps something like "source", "target" and "adj_name". But maybe there's others that are better?

**Comment by dschult** [comment]
In support of the functions in this PR, here are the one-liners that they replace.
 
**First:** ```to_adjacency_dataarray``` with two adjacencies to stack... :
```python
da=xr.concat(
    [
      xr.DataArray(np.expand_dims(nx.adjacency_matrix(G, weight="weight").todense(), axis=-1), 
        dims=["n1","n2","adj_name"], coords={"n1":list(G), "n2":list(G), "name": ['weight_adj_matrix']}),
      xr.DataArray(np.expand_dims(nx.adjacency_matrix(G, weight="capacity").todense(), axis=-1), 
        dims=["n1","n2","adj_name"], coords={"n1":list(G), "n2":list(G), "name": ['capacity_adj_matrix']})
    ],
    dim="name"
) 
```
Notice that the wrapping code around the ```nx.adjacency_matrix(G, weight=...)``` is identical and very specific to naming rows and columns based on the Graph nature of the array. The general DataArray is much more flexible of course, but a graph adjacency version only involves these features -- and it may be helpful for users (and us) not to have people construct their own version of adjacency arrays.

Notice that while it is a one-liner, it's a long one-liner that duplicates the same cruft for each layer of the xarray.
The code in ```to_adjacency_dataarray``` hides much of the one-liner cruft in ```format_adjacency```.

**Second:** ```to_node_dataframe``` with two attributes to stack...:
```python
# setup   nx.set_node_attributes(G, {n: n**2 for n in G}, "n-squared')

df = pd.DataFrame([
    pd.concat([
          pd.Series(d, name=n), 
          pd.Series({"n-fourth": d['n-squared']**2}, name=n)
    ]) 
    for n, d in G.nodes(data=True)
])
```
This is also a one-liner (though I removed some error checking of the ```func``` by inlining.
Again, it would be possible for each user to figure out how to do this. But I think it might be helpful to 
standardize for users a good way to store node information in dataframes.
(Just as we do for ```to_pandas_edgelist``` and ```to_pandas_adjacency```.)  
 
It would take me a long time to come up with these one-liners and I would need to learn a lot about xarray or pandas.
I admit that I would need to learn a fair amount about them just to create the ```funcs``` needed to customize these attributes. But, I think with default functions that simply write out the node/edge attributes on the Graph people can ease into understanding how this works.
